### PR TITLE
Remove copies from TDMProgram and use shots

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
Copies are removed from `TDMProgram` and shots are inserted in the unrolling via `engine.run` instead.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: strawberryfields/engine.py [hunk 1]
```diff
@@ -420,7 +420,10 @@ def run(self, program, *, args=None, compile_options=None, **kwargs):
 
         if not isinstance(program, collections.abc.Sequence) and program.type == "tdm":
             # At this time we do not support lists of tdm programs
-            program.unroll()
+            shots = kwargs.get("shots", 1)
+            program.unroll(shots=shots)
+            # set shots to 1 so that the gaussian backend works
```
Comment (nquesada) [inline strawberryfields/engine.py:425]: Great that you added this comment!

Comment (josh146) [reply strawberryfields/engine.py:425]: Good catch! I have to admit, even with this comment I was still confused for a second 😆 I suggest adding a few more details:

```suggestion
            # Shots >1 for a TDM program simply corresponds to creating
            # multiple copies of the program, and appending them to run sequentially.
            # As a result, we set the backend shots to 1 for the Gaussian backend.
```

File: strawberryfields/engine.py [hunk 2]
```diff
@@ -476,9 +479,9 @@ def run(self, program, *, args=None, compile_options=None, **kwargs):
         if not isinstance(program, collections.abc.Sequence) and program.type == "tdm":
             # At this time we do not support lists of tdm programs
             result._all_samples = reshape_samples(
-                result.all_samples, program.measured_modes, program.N
+                result.all_samples, program.measured_modes, program.N, program.timebins
             )
-            result._samples = np.array(list(result.all_samples.values()))
+            result._samples = np.array(list(result.all_samples.values())).transpose(1, 0, 2)
```
Comment (josh146) [inline strawberryfields/engine.py:484]: How come this transposition is needed? Can it simply be absorbed into the `reshape_samples` function?

File: strawberryfields/tdm/tdmprogram.py [hunk 3]
```diff
@@ -125,21 +121,32 @@ def reshape_samples(all_samples, modes, N):
         all_samples (dict[int, list]): the raw measured samples
         modes (Sequence[int]): the modes that are measured in the circuit
         N (Sequence[int]): the number of concurrent modes per belt/spatial modes
+        timebins (int): the number of timebins/temporal modes in the program per shot
 
     Returns:
-        new_samples (dict[int, list]): the re-shaped samples
+        new_samples (dict[int, list]): the re-shaped samples, where each key correspond to a spatial
+            mode and the values have shape `(shots, timebins)`
```
Comment (josh146) [inline strawberryfields/tdm/tdmprogram.py:128]:
```suggestion
            mode and the values have shape ``(shots, timebins)``
```

File: strawberryfields/tdm/tdmprogram.py [hunk 4]
```diff
@@ -244,12 +251,6 @@ class TDMProgram(sf.Program):
       If the ``p`` variable is not used, the gate argument is assumed to be **constant**
       across all time-bins.
 
-    * ``copies=1`` *(int)*: the number of times to repeat the time-domain program.
-      For example, if a sequence of 10 gate arguments is provided, and ``copies==15``,
-      then there will be a total of :math:`10\times 15` time bins in the time domain
-      program. The sequence corresponding to the gate arguments is repeated ``copies``
-      number of times.
```
Comment (josh146) [inline strawberryfields/tdm/tdmprogram.py:251]: :+1: 🗑️

File: strawberryfields/tdm/tdmprogram.py [hunk 5]
```diff
@@ -447,10 +445,10 @@ def roll(self):
         self.circuit = self.rolled_circuit
         return self
 
-    def unroll(self):
+    def unroll(self, shots=1):
```
Comment (josh146) [inline strawberryfields/tdm/tdmprogram.py:448]: Out of curiosity, how come a default value was added here for shots? Does it make more sense to have `unroll(self, shots)`, so that shots always have to be provided?

File: tests/frontend/test_tdmprogram.py [hunk 6]
```diff
@@ -27,93 +27,81 @@
...
     prog = tdmprogram.TDMProgram(N=2)
-    with prog.context(alpha, phi, theta, copies=copies, shift=shift) as (p, q):
+    with prog.context(alpha, phi, theta, shift=shift) as (p, q):
         ops.Sgate(r, 0) | q[1]
         ops.BSgate(p[0]) | (q[0], q[1])
         ops.Rgate(p[1]) | q[1]
         ops.MeasureHomodyne(p[2]) | q[0]
     eng = sf.Engine("gaussian")
-    result = eng.run(prog)
+    result = eng.run(prog, shots=shots)
```
Comment (josh146) [inline tests/frontend/test_tdmprogram.py:50]: Awesome!

**Review by josh146 — APPROVED** [review]
Left some minor comments, but happy to approve! We shouldn't forget to update the tutorial over the photonics repo, however.

P.S. changelog 😉

**Review by josh146 — CHANGES_REQUESTED** [review]
Actually, I just realised that Travis is not being run - I'm not sure I am comfortable approving without confirmation that all tests pass. We should port the tests to github actions ASAP.

**Comment by nquesada** [comment]
@josh146 : this anyway cannot be merged yet since I still need to modify the tests so that they take into account the new shape of the samples.

**Comment by nquesada** [comment]
Also, once this is merged we will need to update the tutorial.

File: strawberryfields/tdm/tdmprogram.py (hunk 5 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:477]: I think I was just a bit quick adding this here. It makes sense letting it always having to be provided. I'll fix it!

File: strawberryfields/engine.py (hunk 2 above)
Comment (thisac) [reply strawberryfields/engine.py:484]: It's because of the shape change we decided to have for both `all_samples`, `{spatial mode: (shots, timebins)}`, and `samples`, `(shots, spatial modes, timebins)`, along with that `reshape_samples` only actually reshapes and returns `all_samples`. `samples` is created here, outside of the `reshape_samples` function, although I guess this could be move inside of the `reshape_samples` instead, if we want, so that it returns the tuple `(all_samples, samples)` instead.

**Review by josh146 — APPROVED** [review]
Looks good from my end! Small question though - should there be a unittest for `move_vac_modes`? I couldn't spot one.

File: strawberryfields/engine.py [hunk 7]
```diff
@@ -420,7 +420,12 @@ def run(self, program, *, args=None, compile_options=None, **kwargs):
 
         if not isinstance(program, collections.abc.Sequence) and program.type == "tdm":
             # At this time we do not support lists of tdm programs
-            program.unroll()
+            shots = kwargs.get("shots", 1)
+            program.unroll(shots=shots)
+            # Shots >1 for a TDM program simply corresponds to creating
+            # multiple copies of the program, and appending them to run sequentially.
+            # As a result, we set the backend shots to 1 for the Gaussian backend.
+            kwargs["shots"] = 1
```
Comment (antalszava) [inline strawberryfields/engine.py:428]: :100:

File: strawberryfields/tdm/tdmprogram.py [hunk 8]
```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
...
+    dictionary with the shape ``{spatial mode: (shots, timebins)}``. E.g., this unrolled circuit:
+
+    .. code-block:: pycon
+
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 0, spatial mode 0  (sample 0)
+        MeasureHomodyne(0) | (q[2])  # shot 0, timebin 0, spatial mode 2  (sample 1)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 1, spatial mode 0  (sample 2)
+        MeasureHomodyne(0) | (q[1])  # shot 0, timebin 1, spatial mode 2  (sample 3)
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:130]: One question here (think I'm personally just missing something): should the two target modes here be swapped for the measurements? :thinking: 

I'd think that we're shifting with one mode in timebin 1:
1. `MeasureHomodyne(0) | (q[0])` -> `MeasureHomodyne(0) | (q[1])` (spatial mode 0) 
2. `MeasureHomodyne(0) | (q[2])` -> `MeasureHomodyne(0) | (q[0])` (spatial mode 2)

And would have expected:
```
        MeasureHomodyne(0) | (q[1])  # shot 0, timebin 1, spatial mode 0  (sample 2)
        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 1, spatial mode 2  (sample 3)
```

File: strawberryfields/tdm/tdmprogram.py [hunk 9]
```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
...
     if len(set(param_lengths)) != 1:
         raise ValueError("Gate-parameter lists must be of equal length.")
 
-    # check copies
-    if not isinstance(copies, int) or copies < 1:
-        raise TypeError("Number of copies must be a positive integer.")
-
 
-def _get_mode_order(num_of_values, N):
+def _get_mode_order(num_of_values, N, timebins):
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:99]: Would we eventually go with the `timebins` naming in SF (opposed to `temporal_modes`)?

File: strawberryfields/tdm/tdmprogram.py [hunk 10]
```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
...
     """Get the order by which the modes were measured"""
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:100]: (Out of scope:) I find that it would be nice to have an example here and add the arguments, return values to the docstring.

File: strawberryfields/tdm/tdmprogram.py [hunk 11]
```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
...
 
-def reshape_samples(all_samples, modes, N):
+def reshape_samples(all_samples, modes, N, timebins):
     """Reshapes the samples dict so that they have the expected correct shape.
 
     Corrects the :attr:`~.Results.all_samples` dictionary so that the measured modes are the ones
     defined to be measured in the circuit, instead of being spread over a larger
     number of modes due to the mode-shifting occurring in :class:`~.TDMProgram`.
 
+    The function iterates through the unrolled circuit and populates, and returns, a new samples
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:120]:
```suggestion
    The function iterates through samples obtained from the unrolled circuit to populate and return a new samples
```

File: strawberryfields/tdm/tdmprogram.py [hunk 12]
```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 1, spatial mode 0  (sample 2)
+        MeasureHomodyne(0) | (q[1])  # shot 0, timebin 1, spatial mode 2  (sample 3)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 1, timebin 0, spatial mode 0  (sample 4)
+        MeasureHomodyne(0) | (q[2])  # shot 1, timebin 0, spatial mode 2  (sample 5)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 1, timebin 1, spatial mode 0  (sample 6)
+        MeasureHomodyne(0) | (q[1])  # shot 1, timebin 1, spatial mode 2  (sample 7)
+
+    would return the dictionary
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:138]: How about including the equivalent `all_samples` that is produced by the above circuit?

File: strawberryfields/tdm/tdmprogram.py [hunk 13]
```diff
@@ -83,63 +84,136 @@ def validate_measurements(circuit, N):
...
     """
+    # calculate the total number of samples and the order in which they were measured
     num_of_values = len([i for j in all_samples.values() for i in j])
-    mode_order = _get_mode_order(num_of_values, N)
+    mode_order = _get_mode_order(num_of_values, N, timebins)
 
-    # go backwards through all_samples and add them into the correct mode
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
+    t = 0
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:165]: This might make it more descriptive
```suggestion
    timebin_idx = 0
```

File: strawberryfields/tdm/tdmprogram.py [hunk 14]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
         if idx not in new_samples:
-            new_samples[idx] = []
-        new_samples[idx].append(all_samples[m].pop(0))
-    return new_samples
+            new_samples[idx] = [[] for _ in range(timebins)]
+        new_samples[idx][t].append(all_samples[m].pop(0)[0])
+
+        # populate each spatial mode in one timebin before moving on to the next timebin
+        # when each timebin has been filled, move to the next shot
+        if (i + 1) % len(N) == 0:
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:174]: (Minor suggestion): naming the condition could make the operation more descriptive
```suggestion
        # when each timebin has been filled, move to the next shot
        last_mode_in_timebin = (i + 1) % len(N) == 0
        if last_mode_in_timebin:
```

File: strawberryfields/tdm/tdmprogram.py [hunk 15]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
+    t = 0
     for i, m in enumerate(mode_order):
         idx = modes[i % len(N)]
         if idx not in new_samples:
-            new_samples[idx] = []
-        new_samples[idx].append(all_samples[m].pop(0))
-    return new_samples
+            new_samples[idx] = [[] for _ in range(timebins)]
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:169]:
```suggestion
            # create an entry for the new mode with a nested list for each timebin
            new_samples[idx] = [[] for _ in range(timebins)]
```

File: strawberryfields/tdm/tdmprogram.py [hunk 16]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
     num_of_values = len([i for j in all_samples.values() for i in j])
-    mode_order = _get_mode_order(num_of_values, N)
+    mode_order = _get_mode_order(num_of_values, N, timebins)
 
-    # go backwards through all_samples and add them into the correct mode
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
+    t = 0
     for i, m in enumerate(mode_order):
         idx = modes[i % len(N)]
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:167]: I think it would be nice to have this be called `mode_idx`, would make the rest of the function more readable imho.

File: strawberryfields/tdm/tdmprogram.py [hunk 17]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
+    # calculate the total number of samples and the order in which they were measured
     num_of_values = len([i for j in all_samples.values() for i in j])
-    mode_order = _get_mode_order(num_of_values, N)
+    mode_order = _get_mode_order(num_of_values, N, timebins)
 
-    # go backwards through all_samples and add them into the correct mode
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
+    t = 0
     for i, m in enumerate(mode_order):
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:166]: (Minor suggestion)
```suggestion
    for i, mode in enumerate(mode_order):
```

File: strawberryfields/tdm/tdmprogram.py [hunk 18]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
-    return new_samples
+            new_samples[idx] = [[] for _ in range(timebins)]
+        new_samples[idx][t].append(all_samples[m].pop(0)[0])
+
+        # populate each spatial mode in one timebin before moving on to the next timebin
+        # when each timebin has been filled, move to the next shot
+        if (i + 1) % len(N) == 0:
+            t = (t + 1) % timebins
+
+    # transpose each value so that it has shape `(shots, timebins)`
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:177]: Could the mode order creation be adjusted so that transposition would not be needed here? (Such that `new_samples` is populated `(shots, timebins)` by default, might not be a simple change though so feel free to disregard.)

File: strawberryfields/tdm/tdmprogram.py [hunk 19]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
+    return {k: np.array(v).T for k, v in new_samples.items()}
+
+
+def move_vac_modes(samples, N, crop=False):
+    """Post-processing function for TDM samples
+
+    Moves all measured vacuum modes from the first shot of the samples
+    array to the end of the last shot.
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:185]:
```suggestion
    """Post-processing function for TDM samples, moves all measured vacuum modes from the first shot of the samples
    array to the end of the last shot.
```

File: strawberryfields/tdm/tdmprogram.py [hunk 20]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
+
+    Args:
+        samples (ndarray[float]): samples as received from TDMProgram, with the
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:188]:
```suggestion
        samples (ndarray[float]): samples as received from ``TDMProgram``, with the
```

File: strawberryfields/tdm/tdmprogram.py (hunk 8 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:130]: So, in this case this would correspond to `N = [1, 2]`, i.e. one timebin in the first band and 2 timebins in the second. Since the first measurement is in the first band (which only has timebin 0), it stays there:

`MeasureHomodyne(0) | (q[0]) -> MeasureHomodyne(0) | (q[0])`

The second measurement would start on band 2 (which has timebins 1 and 2), and would loop around to timebin 1:

`MeasureHomodyne(0) | (q[2]) -> MeasureHomodyne(0) | (q[1])`

You can check it by running the following:

```pycon
>>> with prog.context([1,2],[4,5]) as (p, q): 
...     BSgate(p[0]) | (q[0], q[1]) 
...     Rgate(p[1]) | q[1] 
...     MeasureHomodyne(p[0]) | q[0] 
...     MeasureHomodyne(p[1]) | q[2] 

>>> eng = sf.Engine("gaussian") 
>>> shots = 2 
>>> result = eng.run(prog, shots=shots)                                    

>>> for c in prog.unrolled_circuit: print(c)
                            
BSgate(1, 0) | (q[0], q[1])
Rgate(4) | (q[1])
MeasureHomodyne(1) | (q[0])
MeasureHomodyne(4) | (q[2])
BSgate(2, 0) | (q[0], q[2])
Rgate(5) | (q[2])
MeasureHomodyne(2) | (q[0])
MeasureHomodyne(5) | (q[1])
```

But this is just one example of the program, and this docstring example doesn't actually tell which circuit we're running (since it's only concerned about the reshaping). :slightly_smiling_face: 

The `...`s would be the other gates in the circuit, that also aren't important for the reshaping.

File: strawberryfields/tdm/tdmprogram.py (hunk 9 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:99]: Don't think we decided upon anything. I feel like, at least, timebins is what is best used inside of `tdmprogram.py` since it's been using that all along, while temporal modes have been used for Blackbird and thus `sf.io` and the device spec also. I think it's fine keeping it like this for, but I guess this is still up for discussion since it might be a bit confusing for users.

File: strawberryfields/tdm/tdmprogram.py (hunk 12 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:138]: What do you mean? The dictionary in the example would be the `all_samples` that's returned by the above circuit (only with `(sample X)` instead of the actual sample value. Or do you mean the earlier `all_samples` that is passed to the `reshape_samples` function? :thinking:

File: strawberryfields/tdm/tdmprogram.py (hunk 18 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:177]: It potentially could, although it wasn't as obvious to me how to do this. I think this is fine, since it would probably take quite a bit of fixing to get it to work the other way (if that's possible without extra overhead anyway at all that is).

File: strawberryfields/tdm/tdmprogram.py [hunk 21]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
+    Returns:
+        ndarray[float]: the post-processed samples
+
+    """
+    num_of_vac_modes = np.max(N) - 1
+    shape = samples.shape
+
+    flat_samples = np.ravel(samples)
+    samples = np.append(flat_samples[num_of_vac_modes:], [0] * num_of_vac_modes)
+    samples = samples.reshape(shape)
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:209]: Wondering if this could be done without having to ravel/reshape the entire samples array :thinking: Maybe using two indices could work? (There wasn't a good alternative that came to mind though. :confused: )

File: strawberryfields/tdm/tdmprogram.py [hunk 22]
```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
...
+
+    if crop:
+        samples = samples[: -num_of_vac_modes // (np.prod(shape[1:]) + 1)]
```
Comment (antalszava) [inline strawberryfields/tdm/tdmprogram.py:212]: Would be nice to break this down into steps (define new variables with descriptive names).

File: tests/frontend/test_tdmprogram.py [hunk 23]
```diff
@@ -146,23 +135,29 @@ def test_epr(self):
...
+        sq_r = 5.0
         c = 2
-        copies = 200
+        shots = 100
 
         # This will generate c EPRstates per copy. I chose c = 4 because it allows us to make 4 EPR pairs per copy that can each be measured in different basis permutations.
-        alpha = [np.pi / 4, 0] * c
-        phi = [0, np.pi / 2] * c
+        alpha = [0, np.pi / 4] * c
+        phi = [np.pi / 2, 0] * c
```
Comment (antalszava) [inline tests/frontend/test_tdmprogram.py:145]: How come these were changed?

File: strawberryfields/tdm/tdmprogram.py (hunk 21 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:209]: I don't think that this is such a big issue though since (correct me if I'm wrong) I believe `np.ravel` and `.reshape` only changes references and doesn't make any copies if the argument is already a NumPy array. So there's quite little overhead even for large sample arrays. Not sure about the appending though.

File: tests/frontend/test_tdmprogram.py (hunk 23 above)
Comment (thisac) [reply tests/frontend/test_tdmprogram.py:145]: @nquesada :question:

File: strawberryfields/engine.py (hunk 2 above)
Comment (antalszava) [reply strawberryfields/engine.py:486]: Worth leaving a comment on this.

**Review by antalszava — COMMENTED** [review]
This looks like a nice change! :100:  Had some comments/questions. Mostly think that further comments and adjustments can be made in the quality of the code, but overall looked good to me.

How would `blackbird` be affected? `copies` is an optional field if I remember correctly, so it would be a "nice to have" adjustment to remove it there?

File: strawberryfields/tdm/tdmprogram.py (hunk 10 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:100]: I would say it's fine leaving this as it is, since the mode order wouldn't be something that visible to the user. It's simply a hidden helper function collecting which modes that were measured in the simulator (nothing to do with the actual modes being returned).

**Comment by thisac** [comment]
> How would blackbird be affected? copies is an optional field if I remember correctly, so it would be a "nice to have" adjustment to remove it there?

@antalszava There's nothing to be changed in Blackbird at all. It doesn't know about `copies`; it simply accepts any keywords. The only changes would be to remove `copies` from Blackbird scripts/templates that have it (but it doesn't harm leaving them either).

File: strawberryfields/tdm/tdmprogram.py (hunk 8 above)
Comment (antalszava) [reply strawberryfields/tdm/tdmprogram.py:130]: Oh got ya, thank you!

When `shots=2`, got the following error though:
```
~/xanadu/strawberryfields/strawberryfields/engine.py in _run_program(self, prog, **kwargs)
    371                         cmd.op, self.backend, kwargs
    372                     )
--> 373                 ) from None
    374 
    375         samples = self._combine_and_sort_samples(samples_dict)

NotImplementedError: The operation MeasureHomodyne(1) has not been implemented in GaussianBackend for the arguments {'shots': 2}.
```

File: strawberryfields/tdm/tdmprogram.py (hunk 9 above)
Comment (antalszava) [reply strawberryfields/tdm/tdmprogram.py:99]: Think this will be confusing in the future in the code base.

File: strawberryfields/tdm/tdmprogram.py (hunk 10 above)
Comment (antalszava) [reply strawberryfields/tdm/tdmprogram.py:100]: Right. Although docstrings are not written entirely for users, I personally found it pretty difficult to parse this function.

File: strawberryfields/tdm/tdmprogram.py (hunk 8 above)
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:130]: Are you running it on the correct branch? That error shouldn't be happening (since, for a TDMProgram, it should set shots to 1 before moving on after unrolling the circuit in `eng.run`).

File: strawberryfields/tdm/tdmprogram.py (hunk 12 above)
Comment (antalszava) [reply strawberryfields/tdm/tdmprogram.py:138]: > do you mean the earlier all_samples that is passed to the reshape_samples function? thinking

Yes, this one.

File: strawberryfields/tdm/tdmprogram.py (hunk 21 above)
Comment (antalszava) [reply strawberryfields/tdm/tdmprogram.py:211]: Hmm, I think you might be right! Checked on an array and there doesn't seem to be considerable overhead added by the ravel/reshape part :+1:

File: strawberryfields/tdm/tdmprogram.py (hunk 8 above)
Comment (antalszava) [reply strawberryfields/tdm/tdmprogram.py:130]: You're right, was my bad! Got confused with branch names

File: tests/frontend/test_tdmprogram.py [hunk 24]
```diff
@@ -640,11 +646,42 @@ def test_tdm_inconsistent_spatial_modes(self):
...
-        prog = singleloop_program(sq_r, alpha, phi, theta, copies)
+        prog = singleloop_program(sq_r, alpha, phi, theta)
         with pytest.raises(sf.program_utils.CircuitError, match="spatial modes, but the device"):
             prog.compile(device=device1, compiler="TD2")
+
+class TestTDMProgramFunctions:
+    """Test functions in the tdmprogram module"""
+
+    @pytest.mark.parametrize("N, crop, expected", [
+        (
```
Comment (antalszava) [inline tests/frontend/test_tdmprogram.py:661]: Worth adding an `N=1`, `crop=True` case to cover the edge case found there.

**Review by antalszava — APPROVED** [review]
Looks good to me! Curious about the parameter changes in the tests, but could be addressed in the TDM HW connection PR too.
